File: 2021_lua/day06.lua
-- Advent Of Code 2021
-- Copyright (C) 2024  Lukas Heindl
-- 
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
-- 
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
-- 
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

local _M = {}

local function parseFile(file)
	local arr = {0,0,0,0,0,0,0,0,0,0}
	for x in file:read("l"):gmatch("(%d+)") do
		arr[x+1] = arr[x+1] + 1
	end
	return arr
end

local function part(file, days)
	--                  0,1,2,3,4,5,6,7,8,9
	local population = parseFile(file)
	-- io.write(string.format("%-3d", 0))
	-- printTable(population)
	for _=1,days do
		local t = table.remove(population, 1)
		population[9] = t
		population[7] = population[7] + t
		-- io.write(string.format("%-3d", i))
		-- printTable(population)
	end
	local sum = 0
	for _,v in pairs(population) do
		sum = sum + v
	end
	return sum
end

function _M.part1(file) return part(file, 80) end
function _M.part2(file) return part(file, 256) end

return _M

File: 2022_Cpp/day06.cpp
/*
Advent Of Code 2020
Copyright (C) 2024  Lukas Heindl

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <cassert>
#include <deque>
#include <fstream>
#include <iostream>
#include <map>
#include <queue>
#include <vector>

#include "aocutils.h"

struct Message
{
	std::string msg;

	int search_start(unsigned int count)
	{
		auto left{msg.begin()};
		auto right{msg.begin()};
		// storee how often which element occurs
		std::map<char, int> eleMap{};

		unsigned int j{0}; // how many characters are being read so far
		// initialize with the first count characters
		for(; j < count; j++, right++) { eleMap[*right]++; }
		// shift window one char to the right each time
		for(; right != msg.end(); right++, left++, j++)
		{
			// count distict elements contained in Map
			if(eleMap.size() == count) return static_cast<int>(j);
			// left char occurs one time less in range, check if it is to be
			// removed
			if(--eleMap[*left] == 0) eleMap.erase(*left);
			// right character occurs one time more in range
			eleMap[*right]++;
		}
		return -1;
	}

	friend std::istream &operator>>(std::istream &is, Message &i)
	{
		// default initialize the most probable moved from object
		i = std::move(Message{});

		std::getline(is, i.msg);
		return is;
	}
};

int main()
{
	// std::ifstream ifs{"../problems/day06.dat.testing"};
	std::ifstream ifs{"../problems/day06.dat"};
	if(ifs.fail()) throw std::runtime_error("File couldn't be opened!");

	Message m{};
	ifs >> m;

	int part1 = m.search_start(4);
	int part2 = m.search_start(14);

	std::cout << "Day 04:" << '\n'
			  << "  Part 1: " << part1 << std::endl
			  << "  Part 2: " << part2 << std::endl;
}
